horreum-backend/src/main/java/io/hyperfoil/tools/horreum/entity/user/UserApiKey.java

```diff
@@ -0,0 +1,164 @@
+        hash = computeHash(keyString());
+        creation = creationDate;
+        revoked = false;
+    }
+
+    public UserService.ApiKeyResponse toResponse() {
+        UserService.ApiKeyResponse response = new UserService.ApiKeyResponse();
+        response.id = id;
+        response.name = name;
+        response.type = type;
+        response.creation = creation;
+        response.access = access;
+        response.isRevoked = revoked;
+        response.toExpiration = toExpiration(LocalDate.now());
+        return response;
+    }
+
+    public boolean isArchived(LocalDate givenDay) {
+        return givenDay.isAfter((access == null ? creation : access).plusDays(active + ARCHIVE_AFTER_DAYS));
+    }
+
+    // calculate the number of days left until expiration (if negative it's the number of days after expiration)
+    private long toExpiration(LocalDate givenDay) {
+        return active - ChronoUnit.DAYS.between(access == null ? creation : access, givenDay);
+    }
+
+    public String keyString() {
+        StringBuilder builder = new StringBuilder(50);
+        builder.append("H");
+        switch (type) {
+            case USER:
+                builder.append("USR");
+                break;
+            default:
+                builder.append("UNK");
+        }
+        builder.append("_");
+        builder.append(randomnessSource.toString().replace("-", "_").toUpperCase()); // keep the dashes for quick validation of key format
+        return builder.toString();
+    }
+
+    // returns the SHA-256 hash of a given key. the hash is what gets sored in the DB, and it's what get compared for authentication
+    private static String computeHash(String key) {
+        try {
+            return Base64.getEncoder().encodeToString(MessageDigest.getInstance("SHA-256").digest(key.getBytes()));
+        } catch (NoSuchAlgorithmException e) {
+            return null; // ignore: SHA-256 should exist
+        }
+    }
+
+    public static Optional<UserApiKey> findOptional(String key) {
+        // validate key structure before computing hash
+        if (key.startsWith("H") && Stream.of(4, 13, 18, 23, 28).allMatch(i -> key.charAt(i) == '_')) {
+            return UserApiKey.<UserApiKey> find("hash", computeHash(key)).firstResultOptional();
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        } else if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        return Objects.equals(this.id, ((UserApiKey) o).id) && Objects.equals(this.hash, ((UserApiKey) o).hash);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, hash);
+    }
+
+    @Override
+    public int compareTo(UserApiKey other) {
+        return Comparator.<UserApiKey, LocalDate> comparing(a -> a.creation).thenComparing(a -> a.id).compare(this, other);
+    }
```

Comment:
It's there so that the keys are always listed in a consistent manner. In the prototype made more sense, so that the user would see keys that were about to expire first.